File: common/src/main/java/juuxel/adorn/client/gui/screen/AbstractConfigScreen.java
package juuxel.adorn.client.gui.screen;

import juuxel.adorn.AdornCommon;
import juuxel.adorn.client.gui.widget.ConfigScreenHeading;
import juuxel.adorn.client.gui.widget.ConfigScreenLabel;
import juuxel.adorn.client.gui.widget.Draggable;
import juuxel.adorn.client.gui.widget.Panel;
import juuxel.adorn.client.gui.widget.ScrollEnvelope;
import juuxel.adorn.config.ConfigManager;
import juuxel.adorn.util.Colors;
import juuxel.adorn.util.Displayable;
import juuxel.adorn.util.PropertyRef;
import juuxel.adorn.util.animation.AnimationEngine;
import juuxel.adorn.util.animation.AnimationTask;
import net.minecraft.client.gl.RenderPipelines;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.screen.NoticeScreen;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.tooltip.Tooltip;
import net.minecraft.client.gui.tooltip.TooltipState;
import net.minecraft.client.gui.widget.CyclingButtonWidget;
import net.minecraft.client.gui.widget.EntryListWidget;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.MathHelper;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class AbstractConfigScreen extends Screen {
    private static final int HEADER_FOOTER_HEIGHT = 33;
    private static final int CONFIG_BUTTON_START_Y = HEADER_FOOTER_HEIGHT + 2;
    public static final int BUTTON_HEIGHT = 20;
    public static final int BUTTON_GAP = 4;
    public static final int BUTTON_SPACING = BUTTON_HEIGHT + BUTTON_GAP;
    protected static final int BACK_BUTTON_Y_FROM_BOTTOM = (HEADER_FOOTER_HEIGHT + BUTTON_HEIGHT) / 2;
    private static final int HEART_SIZE = 12;
    private static final int[] HEART_COLORS = new int[] {
        0xFF_FF0000, // Red
        0xFF_FC8702, // Orange
        0xFF_FFFF00, // Yellow
        0xFF_A7FC58, // Green
        0xFF_2D61FC, // Blue
        0xFF_A002FC, // Purple
        0xFF_58E9FC, // Light blue
        0xFF_FCA1DF, // Pink
    };
    private static final Identifier HEART_TEXTURE = AdornCommon.id("textures/gui/heart.png");
    private static final float MIN_HEART_SPEED = 0.05f;
    private static final float MAX_HEART_SPEED = 1.5f;
    private static final float MAX_HEART_ANGULAR_SPEED = 0.07f;
    private static final int HEART_CHANCE = 65;

    private final Screen parent;

    private final Layout layout;
    private final int backgroundY = HEADER_FOOTER_HEIGHT;
    private int backgroundHeight;
    private Panel mainPanel;

    private final Random random = new Random();
    private final List<Heart> hearts = new ArrayList<>();
    private boolean restartRequired = false;
    private final AnimationEngine animationEngine = new AnimationEngine();

    /** The Y-coordinate of the next config option or heading to be added. */
    protected int nextChildY = CONFIG_BUTTON_START_Y;

    protected AbstractConfigScreen(Text title, Screen parent, Layout layout) {
        super(title);
        this.parent = parent;
        this.layout = layout;
        animationEngine.add(new HeartAnimationTask());
    }

    @Override
    protected void init() {
        nextChildY = CONFIG_BUTTON_START_Y;
        animationEngine.start();

        backgroundHeight = height - backgroundY - HEADER_FOOTER_HEIGHT;
        mainPanel = new Panel();
        initConfigWidgets(mainPanel);
        var sizedPanel = mainPanel.asSized(layout.totalWidth(), nextChildY - backgroundY - BUTTON_GAP);
        addDrawableChild(
            new ScrollEnvelope(
                computeLeftMarginX(),
                backgroundY,
                layout.totalWidth() + ScrollEnvelope.ADDED_WIDTH + 2,
                backgroundHeight,
                sizedPanel,
                animationEngine,
                true
            )
        );
    }

    protected abstract void initConfigWidgets(Panel panel);

    @Override
    public void render(DrawContext context, int mouseX, int mouseY, float delta) {
        super.render(context, mouseX, mouseY, delta);
        context.drawCenteredTextWithShadow(textRenderer, title, width / 2, (HEADER_FOOTER_HEIGHT - textRenderer.fontHeight) / 2, Colors.WHITE);
    }

    @Override
    public void renderBackground(DrawContext context, int mouseX, int mouseY, float tickProgress) {
        super.renderBackground(context, mouseX, mouseY, tickProgress);

        // Draw hearts
        synchronized (hearts) {
            renderHearts(context, tickProgress);
        }

        // Draw darkened background
        var background = client.world == null ? EntryListWidget.MENU_LIST_BACKGROUND_TEXTURE : EntryListWidget.INWORLD_MENU_LIST_BACKGROUND_TEXTURE;
        context.drawTexture(RenderPipelines.GUI_TEXTURED, background, 0, backgroundY, width, backgroundY + backgroundHeight, width, backgroundHeight, 32, 32);

        // Draw headers and footers
        var headerSeparator = client.world == null ? Screen.HEADER_SEPARATOR_TEXTURE : Screen.INWORLD_HEADER_SEPARATOR_TEXTURE;
        var footerSeparator = client.world == null ? Screen.FOOTER_SEPARATOR_TEXTURE : Screen.INWORLD_FOOTER_SEPARATOR_TEXTURE;
        context.drawTexture(RenderPipelines.GUI_TEXTURED, headerSeparator, 0, backgroundY - 2, 0, 0, width, 2, 32, 2);
        context.drawTexture(RenderPipelines.GUI_TEXTURED, footerSeparator, 0, height - HEADER_FOOTER_HEIGHT, 0, 0, width, 2, 32, 2);
    }

    private void renderHearts(DrawContext context, float delta) {
        for (var heart : hearts) {
            var matrices = context.getMatrices();
            matrices.pushMatrix();
            matrices.translate(heart.x, MathHelper.lerp(delta, heart.previousY, heart.y));
            matrices.translate(0.5f * HEART_SIZE, 0.5f * HEART_SIZE);
            var angle = MathHelper.lerp(delta, heart.previousAngle, heart.angle);
            matrices.rotate(angle);
            matrices.translate(-0.5f * HEART_SIZE, -0.5f * HEART_SIZE);
            context.drawTexture(RenderPipelines.GUI_TEXTURED, HEART_TEXTURE, 0, 0, 0f, 0f, HEART_SIZE, HEART_SIZE, 8, 8, 8, 8, heart.color);
            matrices.popMatrix();
        }
    }

    @Override
    public void close() {
        client.setScreen(restartRequired ? new NoticeScreen(
            () -> client.setScreen(parent),
            Text.translatable("gui.adorn.config.restart_required.title"),
            Text.translatable("gui.adorn.config.restart_required.message"),
            Text.translatable("gui.ok"),
            true
        ) : parent);
    }

    @Override
    public void removed() {
        animationEngine.stop();
    }

    private void tickHearts() {
        var iter = hearts.iterator();
        while (iter.hasNext()) {
            var heart = iter.next();

            if (heart.y - HEART_SIZE > height) {
                iter.remove();
            } else {
                heart.move();
            }
        }

        if (random.nextInt(HEART_CHANCE) == 0) {
            int x = random.nextInt(width);
            int color = HEART_COLORS[random.nextInt(HEART_COLORS.length)];
            float speed = random.nextFloat(MIN_HEART_SPEED, MAX_HEART_SPEED);
            float angularSpeed = random.nextFloat(-MAX_HEART_ANGULAR_SPEED, MAX_HEART_ANGULAR_SPEED);
            hearts.add(new Heart(x, -HEART_SIZE, color, speed, angularSpeed));
        }
    }

    private Tooltip createTooltip(PropertyRef<?> property, boolean restartRequired) {
        var text = Text.translatable(getTooltipTranslationKey(property.getName()));
        if (restartRequired) {
            text.append(Text.literal("\n"))
                .append(Text.translatable("gui.adorn.config.requires_restart").formatted(Formatting.ITALIC, Formatting.GOLD));
        }
        return Tooltip.of(text);
    }

    private <T> CyclingButtonWidget<T> createConfigButton(CyclingButtonWidget.Builder<T> builder, PropertyRef<T> property, boolean restartRequired) {
        int x = computeButtonX();
        int y = nextChildY;

        if (layout instanceof Layout.Tabular) {
            builder.omitKeyText();
        }

        return builder
            .tooltip(value -> createTooltip(property, restartRequired))
            .build(x, y, layout.buttonWidth(), BUTTON_HEIGHT, Text.translatable(getOptionTranslationKey(property.getName())), (button, value) -> {
                property.set(value);
                ConfigManager.get().save();

                if (restartRequired) {
                    this.restartRequired = true;
                }
            });
    }

    private int computeLeftMarginX() {
        return (width - layout.totalWidth()) / 2;
    }

    private int computeButtonX() {
        return switch (layout) {
            case Layout.BigButtons(int totalWidth) -> computeLeftMarginX();
            case Layout.Tabular(int totalWidth, int buttonWidth) -> (width + totalWidth) / 2 - buttonWidth;
        };
    }

    private void addConfigLabelIfNeeded(PropertyRef<?> property, boolean restartRequired) {
        if (layout instanceof Layout.Tabular(int totalWidth, int buttonWidth)) {
            var tooltipState = new TooltipState();
            tooltipState.setTooltip(createTooltip(property, restartRequired));
            var label = new ConfigScreenLabel(
                Text.translatable(getOptionTranslationKey(property.getName())),
                tooltipState,
                computeLeftMarginX(),
                nextChildY,
                totalWidth - buttonWidth
            );
            mainPanel.addStandaloneDrawable(label);
        }
    }

    protected void addConfigToggle(PropertyRef<Boolean> property) {
        addConfigToggle(property, false);
    }

    protected void addConfigToggle(PropertyRef<Boolean> property, boolean restartRequired) {
        var button = createConfigButton(
            CyclingButtonWidget.onOffBuilder(property.get()),
            property, restartRequired
        );

        addConfigLabelIfNeeded(property, restartRequired);
        mainPanel.add(button);
        nextChildY += BUTTON_SPACING;
    }

    protected <T extends Displayable> void addConfigButton(PropertyRef<T> property, List<T> values) {
        addConfigButton(property, values, false);
    }

    protected <T extends Displayable> void addConfigButton(PropertyRef<T> property, List<T> values, boolean restartRequired) {
        var button = createConfigButton(
            CyclingButtonWidget.<T>builder(Displayable::getDisplayName, property.get()).values(values),
            property, restartRequired
        );

        addConfigLabelIfNeeded(property, restartRequired);
        mainPanel.add(button);
        nextChildY += BUTTON_SPACING;
    }

    protected void addHeading(Text text) {
        mainPanel.addStandaloneDrawable(new ConfigScreenHeading(text, (width - layout.totalWidth()) / 2, nextChildY, layout.totalWidth()));
        nextChildY += ConfigScreenHeading.HEIGHT;
    }

    protected String getOptionTranslationKey(String name) {
        return "gui.adorn.config.option." + name;
    }

    private String getTooltipTranslationKey(String name) {
        return getOptionTranslationKey(name) + ".description";
    }

    @Override
    public void tick() {
        if (getFocused() instanceof Draggable focused && !isDragging()) {
            focused.stopDragging();
        }
    }

    private static final class Heart {
        private final int x;
        private float y;
        private final int color;
        private final float speed;
        private final float angularSpeed;
        private float previousY;
        private float previousAngle = 0f;
        private float angle = 0f;

        private Heart(int x, float y, int color, float speed, float angularSpeed) {
            this.x = x;
            this.y = y;
            this.color = color;
            this.speed = speed;
            this.angularSpeed = angularSpeed;
            previousY = y;
        }

        private void move() {
            previousY = y;
            y += speed;
            previousAngle = angle;
            angle = (angle + angularSpeed) % MathHelper.TAU;
        }
    }

    private final class HeartAnimationTask implements AnimationTask {
        @Override
        public boolean isAlive() {
            return true;
        }

        @Override
        public void tick() {
            synchronized (hearts) {
                tickHearts();
            }
        }
    }

    public sealed interface Layout {
        int totalWidth();
        int buttonWidth();

        record BigButtons(int buttonWidth) implements Layout {
            @Override
            public int totalWidth() {
                return buttonWidth;
            }
        }

        record Tabular(int totalWidth, int buttonWidth) implements Layout {
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/gui/screen/AbstractConfigScreen.kt
package juuxel.adorn.client.gui.screen

import com.mojang.blaze3d.systems.RenderSystem
import juuxel.adorn.AdornCommon
import juuxel.adorn.client.gui.widget.ConfigScreenHeading
import juuxel.adorn.config.ConfigManager
import juuxel.adorn.util.Colors
import juuxel.adorn.util.Displayable
import juuxel.adorn.util.animation.AnimationEngine
import juuxel.adorn.util.animation.AnimationTask
import juuxel.adorn.util.color
import net.minecraft.client.gui.screen.NoticeScreen
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.CyclingButtonWidget
import net.minecraft.client.render.GameRenderer
import net.minecraft.client.util.OrderableTooltip
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec3f
import kotlin.random.Random
import kotlin.reflect.KMutableProperty

abstract class AbstractConfigScreen(title: Text, private val parent: Screen) : Screen(title) {
    private val random: Random = Random.Default
    private val hearts: MutableList<Heart> = ArrayList()
    private var restartRequired = false
    private val animationEngine = AnimationEngine()

    /** The Y-coordinate of the next config option or heading to be added. */
    protected var nextChildY: Int = CONFIG_BUTTON_START_Y

    init {
        animationEngine.add(HeartAnimationTask())
    }

    override fun init() {
        nextChildY = CONFIG_BUTTON_START_Y
        animationEngine.start()
    }

    override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
        renderBackground(matrices)
        synchronized(hearts) {
            renderHearts(matrices, delta)
        }
        drawCenteredText(matrices, textRenderer, title, width / 2, 20, Colors.WHITE)
        super.render(matrices, mouseX, mouseY, delta)

        for (child in children()) {
            if (child.isMouseOver(mouseX.toDouble(), mouseY.toDouble()) && child is OrderableTooltip) {
                renderOrderedTooltip(matrices, child.orderedTooltip, mouseX, mouseY)
                break
            }
        }
    }

    private fun renderHearts(matrices: MatrixStack, delta: Float) {
        RenderSystem.setShader(GameRenderer::getPositionTexColorShader)
        RenderSystem.setShaderTexture(0, HEART_TEXTURE)

        for (heart in hearts) {
            RenderSystem.setShaderColor(Colors.redOf(heart.color), Colors.greenOf(heart.color), Colors.blueOf(heart.color), 1f)
            matrices.push()
            matrices.translate(heart.x.toDouble(), MathHelper.lerp(delta.toDouble(), heart.previousY, heart.y), 0.0)
            matrices.translate(HEART_SIZE.toDouble() / 2, HEART_SIZE.toDouble() / 2, 0.0)
            matrices.multiply(Vec3f.POSITIVE_Z.getRadialQuaternion(heart.angle.toFloat()))
            matrices.translate(-HEART_SIZE.toDouble() / 2, -HEART_SIZE.toDouble() / 2, 0.0)
            drawTexture(matrices, 0, 0, HEART_SIZE, HEART_SIZE, 0f, 0f, 8, 8, 8, 8)
            matrices.pop()
        }
    }

    override fun close() {
        client!!.setScreen(
            if (restartRequired) {
                NoticeScreen(
                    { client!!.setScreen(parent) },
                    Text.translatable("gui.adorn.config.restart_required.title"),
                    Text.translatable("gui.adorn.config.restart_required.message"),
                    Text.translatable("gui.ok"),
                    true
                )
            } else {
                parent
            }
        )
    }

    override fun removed() {
        animationEngine.stop()
    }

    private fun tickHearts() {
        val iter = hearts.iterator()
        while (iter.hasNext()) {
            val heart = iter.next()

            if (heart.y - HEART_SIZE > height) {
                iter.remove()
            } else {
                heart.move()
            }
        }

        if (random.nextInt(HEART_CHANCE) == 0) {
            val x = random.nextInt(width)
            val color = HEART_COLORS.random(random)
            val speed = random.nextDouble(MIN_HEART_SPEED, MAX_HEART_SPEED)
            val angularSpeed = random.nextDouble(-MAX_HEART_ANGULAR_SPEED, MAX_HEART_ANGULAR_SPEED)
            hearts += Heart(x, -HEART_SIZE.toDouble(), color, speed, angularSpeed)
        }
    }

    private fun wrapTooltipLines(text: Text) =
        textRenderer.wrapLines(text, 200)

    private fun <T> createConfigButton(
        builder: CyclingButtonWidget.Builder<T>, x: Int, y: Int, width: Int, property: KMutableProperty<T>, restartRequired: Boolean
    ) = builder.tooltip {
        buildList {
            this += wrapTooltipLines(Text.translatable(getTooltipTranslationKey(property.name)))

            if (restartRequired) {
                this += wrapTooltipLines(
                    Text.translatable("gui.adorn.config.requires_restart").formatted(Formatting.ITALIC, Formatting.GOLD)
                )
            }
        }
    }.build(x, y, width, BUTTON_HEIGHT, Text.translatable(getOptionTranslationKey(property.name))) { _, value ->
        property.setter.call(value)
        ConfigManager.INSTANCE.save()

        if (restartRequired) {
            this.restartRequired = true
        }
    }

    protected fun addConfigToggle(
        width: Int, property: KMutableProperty<Boolean>, restartRequired: Boolean = false
    ) {
        val button = createConfigButton(
            CyclingButtonWidget.onOffBuilder(property.getter.call()),
            (this.width - width) / 2, nextChildY, width, property, restartRequired
        )
        addDrawableChild(button)
        nextChildY += BUTTON_SPACING
    }

    protected fun <T : Displayable> addConfigButton(
        width: Int, property: KMutableProperty<T>, values: List<T>, restartRequired: Boolean = false
    ) {
        val button = createConfigButton(
            CyclingButtonWidget.builder<T> { it.displayName }.values(values).initially(property.getter.call()),
            (this.width - width) / 2, nextChildY, width, property, restartRequired
        )
        addDrawableChild(button)
        nextChildY += BUTTON_SPACING
    }

    protected fun addHeading(text: Text, width: Int) {
        addDrawable(ConfigScreenHeading(text, (this.width - width) / 2, nextChildY, width))
        nextChildY += ConfigScreenHeading.HEIGHT
    }

    protected open fun getOptionTranslationKey(name: String): String =
        "gui.adorn.config.option.$name"

    private fun getTooltipTranslationKey(name: String): String =
        "${getOptionTranslationKey(name)}.description"

    companion object {
        private const val CONFIG_BUTTON_START_Y = 40
        const val BUTTON_HEIGHT = 20
        const val BUTTON_GAP = 4
        const val BUTTON_SPACING = BUTTON_HEIGHT + BUTTON_GAP
        private const val HEART_SIZE = 12
        private val HEART_COLORS: List<Int> = listOf(
            color(0xFF0000), // Red
            color(0xFC8702), // Orange
            color(0xFFFF00), // Yellow
            color(0xA7FC58), // Green
            color(0x2D61FC), // Blue
            color(0xA002FC), // Purple
            color(0x58E9FC), // Light blue
            color(0xFCA1DF), // Pink
        )
        private val HEART_TEXTURE = AdornCommon.id("textures/gui/heart.png")
        private const val MIN_HEART_SPEED = 0.05
        private const val MAX_HEART_SPEED = 1.5
        private const val MAX_HEART_ANGULAR_SPEED = 0.07
        private const val HEART_CHANCE = 65
    }

    private class Heart(val x: Int, var y: Double, val color: Int, val speed: Double, val angularSpeed: Double) {
        var previousY: Double = y
        var previousAngle: Double = 0.0
        var angle: Double = 0.0

        fun move() {
            previousY = y
            y += speed
            previousAngle = angle
            angle = (angle + angularSpeed) % MathHelper.TAU
        }
    }

    private inner class HeartAnimationTask : AnimationTask {
        override fun isAlive(): Boolean = true
        override fun tick() {
            synchronized(hearts) {
                tickHearts()
            }
        }
    }
}
